src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -1355,18 +1358,50 @@ MulticopterPositionControl::task_main()
 					_vel_sp(1) = (_pos_sp(1) - _pos(1)) * _params.pos_p(1);
 				}
 
-				if (_run_alt_control) {
-					_vel_sp(2) = (_pos_sp(2) - _pos(2)) * _params.pos_p(2);
-				}
+				// do not go slower than the follow target velocity when position tracking is active
+
+                if(_pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_FOLLOW_TARGET &&
+                   _pos_sp_triplet.current.velocity_valid) {
+
+                    math::Vector<3> ft_vel(_pos_sp_triplet.current.vx, _pos_sp_triplet.current.vy, 0);
+
+                    float cos_ratio = (ft_vel*_vel_sp)/(ft_vel.length()*_vel_sp.length());
+
+                    // only override velocity set points when uav is traveling in same direction as target and vector component
+                    // is greater than calculated position set point velocity component
 
-				/* make sure velocity setpoint is saturated in xy*/
-				float vel_norm_xy = sqrtf(_vel_sp(0) * _vel_sp(0) +
-							  _vel_sp(1) * _vel_sp(1));
+                    if(cos_ratio > 0) {
+                        ft_vel *= (cos_ratio);
+                        // min speed a little faster than target vel
+                        ft_vel += ft_vel.normalized()*1.5f;
+                    } else {
+                        ft_vel.zero();
+                    }
 
-				if (vel_norm_xy > _params.vel_max(0)) {
-					/* note assumes vel_max(0) == vel_max(1) */
-					_vel_sp(0) = _vel_sp(0) * _params.vel_max(0) / vel_norm_xy;
-					_vel_sp(1) = _vel_sp(1) * _params.vel_max(1) / vel_norm_xy;
+                    _vel_sp(0) = fabs(ft_vel(0)) > fabs(_vel_sp(0)) ? ft_vel(0) : _vel_sp(0);
+                    _vel_sp(1) = fabs(ft_vel(1)) > fabs(_vel_sp(1)) ? ft_vel(1) : _vel_sp(1);
+                }
+
+                // Ignore position control calculated vel set points
```

Comment:
The above code uses a combination of overriding the velocity set point when it is appropriate while still using the position set point when it's not so I can take advantage of the velocity braking code as the position set point error delta gets smaller.  The code below overrides the velocity set point at all times and ignores the position set point.  I was going to make a new set point type for the code below but saw that the position set point type already defined a set point type for velocity only, it just wasn't being used in the code anywhere. I used both of these set point types for a smoother transition between following using position and velocity. Let me know if you still think I need to fix something.

https://github.com/PX4/Firmware/pull/3814/files?diff=split#diff-7a824e3d1e9a5420071c8c9bbe27b8e7R366